### PR TITLE
[CP Stg] Replace ReportScreenIDSetter with useLastAccessedReportID

### PR DESCRIPTION
**Pull request by roryabraham**
### Details
Some discussion in slack: https://expensify.slack.com/archives/C01GTK53T8Q/p1719427166028799

TL;DR:

- react-navigation assumes that you'll use the `navigation` prop in response to user actions. We are using `navigation.setParams` to set initial params for a screen. The preferred way to do that is with `initialParams`.
- I'm not 100% sure it's necessary, but I _think_ it might be possible that there's a race condition between `useLastAccessedReportID` (i.e: `useOnyx`/`useSyncExternalStore`) and react-navigation mounting the navigator. So in some cases the `ReportScreen` might fully mount before `useLastAccessedReportID` returns the value computed from Onyx. To cover our bases there, we add a `useEffect` in `ReportScreen` that sets the `reportID` route param to `lastAccessedReportID` if it's not present
- Removed the unnecessary `ReportScreenWrapper` and `ReportScreenIDSetter` (no-op) components from the view hierarchy.

Less TL;DR version:

* Previously, we were using `withOnyx` in `ReportScreenIDSetter` to calculate the `lastAccessedReportID`, and if we did not have a reportID in the URL, we would use `navigation.setParams` to set the reportID to `lastAccessedReportID`. The final effect was that we "re-route" from `r/` to `r/$lastAccessedReportID`
* Little did we know, this only worked because the render-blocking behavior of `withOnyx` delayed the call to `navigation.setParams` enough that it gave time for the navigation tree to finish mounting
* When we used `useOnyx` (which uses `useSyncExternalStore` under the hood), `ReportScreenIDSetter` became much faster and the render-blocking behavior was removed. As a result `navigation.setParams` was being called before the navigation tree was finished mounting. This did not work as expected because the `react-navigation` docs assume that you'll be calling `navigate` and/or `setParams` in response to user actions, and after the navigation tree had finished mounting.
* The solution to that problem was to:
    * move the logic that calculates `lastAccessedReportID` up the tree to `AuthScreens`
    * pass `{reportID: lastAccessedReportID}` to `initialParams` for the `ReportScreen`, if there wasn't already a reportID in the route
* However, that revealed another interesting side-effect of our previous design. `ReportScreenIDSetter` was a no-op component that only contained a `useEffect` and returned null. Therefore, we didn't notice (and perhaps it didn't matter) that it would re-render any time _any_ of its Onyx data changes (any report, reportMetadata, policy, etc...). When we moved this to `AuthScreens`, it created problems. Not only would this likely have been a performance regression to re-render `AuthScreens` so aggressively, but also it put us in an infinite loop (in the test environment, anyways...)
    1. AuthScreens mounts
    2. ReportScreen mounts and sets `reportMetadata.lastVisitTime` for the report
    3. (new) AuthScreens re-renders because `reportMetadata` changed
    4. `ReportScreen` remounts because the navigator it lives in re-rendered
    5. it sets `reportMetadata.lastVisitTime`, and we go back to step C
* To fix that issue, I ended up coming up with custom solution. It's a new pattern, but it worked well for our needs in this case... It uses `useSyncExternalStore` and `Onyx.connect` to subscribe to all the keys needed to compute `lastAccessedReportID`, and then computes the `lastAccessedReportID` property in `getSnapshot`. The result is that even though we still access and respond to all that big Onyx data (reports, reportMetadata, policies, etc...), `useLastAccessedReportID` will only trigger a re-render if `lastAccessedReportID` changes

### Fixed Issues
$ https://github.com/Expensify/App/issues/44442
PROPOSAL: https://expensify.slack.com/archives/C01GTK53T8Q/p1719506066435689?thread_ts=1719427166.028799&cid=C01GTK53T8Q

### Tests / QA Steps
1. Open any report
1. Open the `New Chat` page
1. Select a few members of a group, then press `Next`
1. Refresh the page.
1. Verify that the report you had open loads in the background.

- [x] Verify that no errors appear in the JS console

### Offline tests
None - can't refresh the page offline because our site isn't set up with ServiceWorkers.

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/47436092/ddb4850d-3b83-4062-a9c4-528bf4f42d6f

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by c3024** [comment]
I am seeing a continuous loading of the page on iOS

https://github.com/Expensify/App/assets/102477862/0844746c-9a5b-4e61-a52e-d0020f4af426

File: src/hooks/useLastAccessedReportID.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * Get the last accessed reportID.
+ */
+export default function useLastAccessedReportID(shouldOpenOnAdminRoom: boolean) {
+    const {canUseDefaultRooms} = usePermissions();
+    const {activeWorkspaceID} = useActiveWorkspace();
+
+    const [reports] = useOnyx(ONYXKEYS.COLLECTION.REPORT, {allowStaleData: true});
+    const [policies] = useOnyx(ONYXKEYS.COLLECTION.POLICY, {allowStaleData: true});
+    const [isFirstTimeNewExpensifyUser = false] = useOnyx(ONYXKEYS.NVP_IS_FIRST_TIME_NEW_EXPENSIFY_USER);
+    const [reportMetadata] = useOnyx(ONYXKEYS.COLLECTION.REPORT_METADATA, {allowStaleData: true});
```
Comment (roryabraham) [inline src/hooks/useLastAccessedReportID.ts:19]: The failing test appears to be caused by an infinite loop. By process of elimination, I have traced the source to this line

File: src/hooks/useLastAccessedReportID.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            policies,
+            isFirstTimeNewExpensifyUser,
+            shouldOpenOnAdminRoom,
+            reportMetadata,
+            activeWorkspaceID,
+            policyMemberAccountIDs,
+        )?.reportID;
+    }, [activeWorkspaceID, canUseDefaultRooms, shouldOpenOnAdminRoom]);
+
+    // We need access to all the data from these useOnyx calls, but we don't want to re-render the consuming component
```
Comment (mountiny) [inline src/hooks/useLastAccessedReportID.ts:83]:
```suggestion
    // We need access to all the data from these Onyx.connect calls, but we don't want to re-render the consuming component
```

File: src/hooks/useLastAccessedReportID.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    // unless the derived value (lastAccessedReportID) changes. To address these, we'll wrap everything with
```
Comment (mountiny) [inline src/hooks/useLastAccessedReportID.ts:84]: I guess it could be omitted but I think its better to leave it in the comment too
```suggestion
    // unless the derived value (lastAccessedReportID) changes. To address these, we'll wrap everything with useSyncExternalStore
```

File: src/pages/home/ReportScreen.tsx [hunk 4]
```diff
@@ -161,6 +163,29 @@ function ReportScreen({
     const isLoadingReportOnyx = isLoadingOnyxValue(reportResult);
     const permissions = useDeepCompareRef(reportOnyx?.permissions);
 
+    // check if there's a reportID in the route. If not, set it to the last accessed reportID
```
Comment (mountiny) [inline src/pages/home/ReportScreen.tsx:166]:
```suggestion
    // Check if there's a reportID in the route. If not, set it to the last accessed reportID
```

**Review by mountiny — DISMISSED** [review]
Looks good, only NAB comments

**Comment by c3024** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

File: src/pages/home/ReportScreen.tsx [hunk 5]
```diff
@@ -161,6 +163,29 @@ function ReportScreen({
...
+    const lastAccessedReportID = useLastAccessedReportID(!!route.params.openOnAdminRoom);
+    useEffect(() => {
+        // Don't update if there is a reportID in the params already
+        if (route.params.reportID) {
+            const reportActionID = route?.params?.reportActionID;
+            const regexValidReportActionID = new RegExp(/^\d*$/);
```
Comment (luacmartins) [inline src/pages/home/ReportScreen.tsx:172]: NAB we could use ValidationUtils.isNumeric
```suggestion
            const isValidReportActionID = ValidationUtils.isNumeric(reportActionID);
```

Comment (roryabraham) [reply src/pages/home/ReportScreen.tsx:172]: This was existing code, but in for a penny in for a pound I suppose 😉

**Review by mountiny — APPROVED** [review]

**Comment by c3024** [comment]
On refreshing the page with `Start Chat` RHP open, the center pane report keeps loading with a skeleton loader.


https://github.com/Expensify/App/assets/102477862/69dd5b40-e567-456c-b518-2c5e913b2c4f

It works fine in the adhoc build.

**Comment by c3024** [comment]
@mountiny can you please run the build again now that all changes (hopefully) are done?

**Review by luacmartins — APPROVED** [review]
LGTM

**Review by c3024 — APPROVED** [review]
Tests well!

File: src/pages/home/ReportScreen.tsx [hunk 6]
```diff
@@ -161,6 +164,29 @@ function ReportScreen({
...
 
+    // Check if there's a reportID in the route. If not, set it to the last accessed reportID
+    const lastAccessedReportID = useLastAccessedReportID(!!route.params.openOnAdminRoom);
+    useEffect(() => {
+        // Don't update if there is a reportID in the params already
+        if (route.params.reportID) {
+            const reportActionID = route?.params?.reportActionID;
+            const isValidReportActionID = ValidationUtils.isNumeric(reportActionID);
+            if (reportActionID && !isValidReportActionID) {
+                navigation.setParams({reportActionID: ''});
```
Comment (allroundexperts) [inline src/pages/home/ReportScreen.tsx:175]: We should have waited for the navigation to get initialised. Not doing so caused https://github.com/Expensify/App/issues/63347
